### ISSUE TITLE
Shortcuts using Alt are not recognized

### ISSUE THREAD

**Issue by rien333** [kovidgoyal/kitty#3446]
**Describe the bug**
When compiled from git, kitty doesn't seem to recognize shortcuts performed using the left Alt key. 

If I do perform such a shortcut, it's as if I hadn't pressed the Alt key at all — the alphanumeric key is simply inserted verbatim, instead of a shortcut being triggered (so <kbd>Alt + f</kbd> would just insert "f").

**To Reproduce**
1. `kitty --config None bash`
2. Press <kbd>Alt+f</kbd>

**Expected behavior**
Kitty should process <kbd>Alt+f</kbd> as "move forward word" (in e.g. bash/fish), instead of simply inserting an "f" character.

**Environment details**

The bug doesn't occur on the latest stable version (0.19.3), but was probably introduced in a commit somewhere after that.

OS: Linux

```console
$ kitty --debug-config
kitty 0.19.3 (7e78f30739) created by Kovid Goyal
Linux bummer 5.11.11-arch1-1-custom #1 SMP PREEMPT Wed, 31 Mar 2021 07:41:27 +0000 x86_64
Arch Linux \r (\l)
LSB_VERSION=1.4
DISTRIB_ID=Arch
DISTRIB_RELEASE=rolling
DISTRIB_DESCRIPTION="Arch Linux"
Loaded config files: /home/rw/.config/kitty/kitty.conf
Running under: X11
```

**Additional context**

This is the output if press <kbd>Alt+f</kbd> when I launch kitty with the `--debug-keyboard` flag:

```
Loading new XKB keymaps
Modifier indices alt:-1 super:6 hyper:5 meta:-1 numlock:-1
Press xkb_keycode: 0x40 clean_sym: Alt_L composed_sym: Alt_L mods: none glfw_key: 57442 (LEFT_ALT) xkb_key: 65513 (Alt_L)
on_key_input: glfw key: 57442 native_code: 0xffe9 action: PRESS mods: 0x0 text: '' state: 0 ignoring as keyboard mode does not allow repeat events
Press xkb_keycode: 0x29 clean_sym: f composed_sym: f active_unknown_mods: Mod1 text: f mods: none glfw_key: 102 (f) xkb_key: 102 (f)
on_key_input: glfw key: 102 native_code: 0x66 action: PRESS mods: 0x0 text: 'f' state: 0 sent text to child
Release xkb_keycode: 0x29 clean_sym: f mods: none glfw_key: 102 (f) xkb_key: 102 (f)
on_key_input: glfw key: 102 native_code: 0x66 action: RELEASE mods: 0x0 text: '' state: 0 ignoring as keyboard mode does not allow release events
Release xkb_keycode: 0x40 clean_sym: Alt_L mods: none glfw_key: 57442 (LEFT_ALT) xkb_key: 65513 (Alt_L)
on_key_input: glfw key: 57442 native_code: 0xffe9 action: RELEASE mods: 0x0 text: '' state: 0 ignoring as keyboard mode does not allow release events
```

**Comment by kovidgoyal** [comment]
I cannot reproduce, running

kitty --config NONE showkey -a

and pressing alt+f

generates the output

```
^[f 	27 0033 0x1b
 	    102 0146 0x66
```

as expected.

**Comment by rien333** [comment]
Bummer.  For me, `showkey` shows a plain f when I type <kbd>Alt+f</kbd>, but looks normal if I type <kbd>Ctrl+f</kbd>. May try to bisect this tomorrow. 

```
f 	102 0146 0x66
^F 	 6 0006 0x06
```

**Comment by kovidgoyal** [comment]
Use --debug-keyboard its likely a result of #3430 ping @orki

**Comment by rien333** [comment]
> Use --debug-keyboard

FTR: I did do so (see the bottom of my original post)

**Comment by kovidgoyal** [comment]
Need the full output not just the last lines, near the top there will be something about modifier indices.

**Comment by rien333** [comment]
Edited, sorry.

**Comment by kovidgoyal** [comment]
Yeah as I thought, the new code is not recognizing some mods on your system. Lets wait for @orki

**Comment by orki** [comment]
Not clear to me that this is the right fix. If `numLock` is not assigned to any modifier, then the new code will appropriate one, when it shouldn't. @rien333 Do you have a custom xkb format?

**Comment by orki** [comment]
@rien333 Could you run `xkbcomp $DISPLAY debug.xkb` and attach the resulting file called `debug.xkb`?

**Comment by kovidgoyal** [comment]
On Tue, Apr 06, 2021 at 12:31:33PM -0700, orki wrote:
> Not clear to me that this is the right fix. If `numLock` is not assigned to any modifier, then the new code will appropriate one, when it shouldn't. @rien333 Do you have a custom xkb format?

Well, it's pretty evident that the new code is not correct either, since
it fails to assign even basic mods like alt on some systems. I prefer to
be in a situation where the basic mods work and more exotic ones fail. 

I will note that no one else has reported issues with the existing code
and as far as I know all the major toolkits such as Qt just use the xkb
function without needing to use XkbGetMap. For example see
qxkbcommon.c in the Qt source code.

That said I am not wedded to my patch.

**Comment by orki** [comment]
That's fair. I'd really like to figure out what's going on with OP's configuration. I hope they return with their xkb keymap.

**Comment by kovidgoyal** [comment]
Yes, understanding the root cause would of course be preferable.

**Comment by orki** [comment]
It appears that I missed `xkb_state_update_mask` which converts from X modifiers to xkb modifiers, and which should work in wayland too. I'll get to it this weekend.
That said, I still want to figure out why the OP's configuration does not work with the current code.

**Comment by kovidgoyal** [comment]
Isnt xkb_state_update_mask already called in glfw_xkb_update_modifiers() which in turn is called in response to XkbStateNotify events.

**Comment by rien333** [comment]
Sorry, felt pretty ill yesterday. I'll do some more debugging in a bit.

**Comment by rien333** [comment]
I found that,  in order to reproduce my issue,  you need my .Xmodmap. My Xmodmap looks okay to me; though it may well be that I've misconfigured something (but then again, I've been using this config for about two years) 

```
clear lock

clear      mod1
clear      mod2
clear      mod3
clear      mod4

keycode 66 = Hyper_L
keycode 135 = Hyper_L

add mod1 = Alt_L Meta_L
add mod2 = Alt_R
add mod3 = Hyper_L
add mod4 = Super_L Super_R
```

@kovidgoyal note that ac3080f0e25f11252d47797db0044661d7554e52 fixed my issue.

@orki I've attached my debug.xkb (had to change the extention to .txt though). Thanks for the help!
[debug.txt](https://github.com/kovidgoyal/kitty/files/6271021/debug.txt)

**Comment by orki** [comment]
This is very helpful @rien333. An assumption that I made in my implementation was that two real X modifiers (Mod1 and Mod2 in your case) would not represent the same virtual modifier (Alt in your case), but your setup breaks the assumption. Let me think about a better solution. Whatever the solution is, the assumption in code (prior to my patch) that `altIndx` and `altMask` have a fixed relationship will need to be broken.

**Comment by orki** [comment]
I was able to reproduce the problem; as explained in https://github.com/kovidgoyal/kitty/issues/3446#issuecomment-814964523, the issue was multiple real modifiers mapping to the same virtual modifier. The attached patch fixes the problem.
[0001-Handle-multiple-real-modifiers-mapping-to-the-same-v.txt](https://github.com/kovidgoyal/kitty/files/6275655/0001-Handle-multiple-real-modifiers-mapping-to-the-same-v.txt)

**Comment by orki** [comment]
Replying to https://github.com/kovidgoyal/kitty/issues/3446#issuecomment-814568467:
Qt follows exactly the same algorithm as in pull request #3430 in their `QXcbKeyboard::updateVModToRModMapping` with the main difference being the use of `xcb_xkb_get_map`, the xcb wrapper of `XKBGetMap`:
```
    auto map_reply = Q_XCB_REPLY(xcb_xkb_get_map,
                                 xcb_connection(),
                                 XCB_XKB_ID_USE_CORE_KBD,
                                 XCB_XKB_MAP_PART_VIRTUAL_MODS,
                                 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0);
  // some error checking and unpacking code elided
    uint vmod_mask, bit;
    // the virtual modifiers mask for which a set of corresponding
    // real modifiers is to be returned
    vmod_mask = map_reply->virtualMods;
    int count = 0;

    for (bit = 1; vmod_mask; bit <<= 1) {
        uint modmap;

        if (!(vmod_mask & bit))
            continue;

        vmod_mask &= ~bit;
        // real modifier bindings for the specified virtual modifiers
        modmap = map.vmods_rtrn[count];
        count++;

        if (vmod_masks.alt == bit)
            rmod_masks.alt = modmap;
        else if (vmod_masks.meta == bit)
            rmod_masks.meta = modmap;
        else if (vmod_masks.altgr == bit)
            rmod_masks.altgr = modmap;
        else if (vmod_masks.super == bit)
            rmod_masks.super = modmap;
        else if (vmod_masks.hyper == bit)
            rmod_masks.hyper = modmap;
    }
```
If there is a reason to prefer xcb connection to calling X11 directly, we can probably do the same.

**Comment by kovidgoyal** [comment]
@orki please send a new PR and I will review.

**Comment by orki** [comment]
@rien333 Does it work for you now with the fix from PR #3452?

**Comment by rien333** [comment]
Yes, it does, thanks again. Sorry for not coming back to this.
